test/component/docker/docker.go

```diff
@@ -0,0 +1,84 @@
+// +build component
+
+package docker
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/opencontainers/go-digest"
+	"io"
+	"net/url"
+	"os"
+)
+
+type RegistryConfig struct {
+	URL      *url.URL
+	Username string
+	Password string
+}
+
+func (c RegistryConfig) GetRegistryAuth() (auth string, err error) {
+	authConfig := types.AuthConfig{
+		Username: c.Username,
+		Password: c.Password,
+	}
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return "", err
+	}
+	auth = base64.URLEncoding.EncodeToString(encodedJSON)
+	return
+}
+
+func (c RegistryConfig) GetBasicAuthorization() string {
+	return fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%s:%s", c.Username, c.Password))))
+}
+
+// ReplicateImage tags the given imageRef and pushes it to the given dest registry.
+func ReplicateImage(imageRef string, dest RegistryConfig) (d digest.Digest, err error) {
+	ctx := context.Background()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return
+	}
+	pullOut, err := cli.ImagePull(ctx, imageRef, types.ImagePullOptions{})
+	defer func() {
+		_ = pullOut.Close()
+	}()
+
+	_, err = io.Copy(os.Stdout, pullOut)
```

Comment:
This breaks stdout like the following.
```
{"status":"Preparing","progressDetail":{},"id":"273420d2bcff"}
{"status":"Preparing","progressDetail":{},"id":"6189abe095d5"}
{"status":"Preparing","progressDetail":{},"id":"f47163e8de57"}
{"status":"Preparing","progressDetail":{},"id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":38400,"total":1933288},"progress":"[\u003e                                                  ]   38.4kB/1.933MB","id":"6189abe095d5"}
{"status":"Pushing","progressDetail":{"current":41472,"total":1825650},"progress":"[=\u003e                                                 ]  41.47kB/1.826MB","id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":539648,"total":176851124},"progress":"[\u003e                                                  ]  539.6kB/176.9MB","id":"273420d2bcff"}
{"status":"Pushing","progressDetail":{"current":524288,"total":1933288},"progress":"[=============\u003e                                     ]  524.3kB/1.933MB","id":"6189abe095d5"}
{"status":"Pushing","progressDetail":{"current":282112,"total":1825650},"progress":"[=======\u003e                                           ]  282.1kB/1.826MB","id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":1802240,"total":1933288},"progress":"[==============================================\u003e    ]  1.802MB/1.933MB","id":"6189abe095d5"}
{"status":"Pushing","progressDetail":{"current":1966080,"total":1933288},"progress":"[==================================================\u003e]  1.966MB","id":"6189abe095d5"}
{"status":"Pushing","progressDetail":{"current":160968,"total":15073004},"progress":"[\u003e                                                  ]    161kB/15.07MB","id":"f47163e8de57"}
{"status":"Pushing","progressDetail":{"current":562679,"total":1825650},"progress":"[===============\u003e                                   ]  562.7kB/1.826MB","id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":2144256,"total":176851124},"progress":"[\u003e                                                  ]  2.144MB/176.9MB","id":"273420d2bcff"}
{"status":"Pushing","progressDetail":{"current":1318400,"total":15073004},"progress":"[====\u003e                                              ]  1.318MB/15.07MB","id":"f47163e8de57"}
{"status":"Pushing","progressDetail":{"current":729088,"total":1825650},"progress":"[===================\u003e                               ]  729.1kB/1.826MB","id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":3239424,"total":176851124},"progress":"[\u003e                                                  ]  3.239MB/176.9MB","id":"273420d2bcff"}
{"status":"Pushing","progressDetail":{"current":2501632,"total":15073004},"progress":"[========\u003e                                          ]  2.502MB/15.07MB","id":"f47163e8de57"}
{"status":"Pushing","progressDetail":{"current":767047,"total":1825650},"progress":"[=====================\u003e                             ]    767kB/1.826MB","id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":4853760,"total":176851124},"progress":"[=\u003e                                                 ]  4.854MB/176.9MB","id":"273420d2bcff"}
{"status":"Pushing","progressDetail":{"current":3484648,"total":15073004},"progress":"[===========\u003e                                       ]  3.485MB/15.07MB","id":"f47163e8de57"}
{"status":"Pushing","progressDetail":{"current":804650,"total":1825650},"progress":"[======================\u003e                            ]  804.6kB/1.826MB","id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":4768936,"total":15073004},"progress":"[===============\u003e                                   ]  4.769MB/15.07MB","id":"f47163e8de57"}
{"status":"Pushing","progressDetail":{"current":5952000,"total":176851124},"progress":"[=\u003e                                                 ]  5.952MB/176.9MB","id":"273420d2bcff"}
{"status":"Pushing","progressDetail":{"current":993109,"total":1825650},"progress":"[===========================\u003e                       ]  993.1kB/1.826MB","id":"42a3027eaac1"}
{"status":"Pushing","progressDetail":{"current":6745088,"total":15073004},"progress":"[======================\u003e                            ]  6.745MB/15.07MB","id":"f47163e8de57"}
```

If this output is not important, we can use `ioutil.Discard`.
```suggestion
	_, err = io.Copy(ioutil.Discard, pullOut)
```

Reply:
Thanks. `Discar` writer makes much more sense in this case.

Reply:
Done